R/age_model_class.R

```diff
@@ -0,0 +1,222 @@
+#' Defines an age-structured simple SEIR model.
+#' Slots of this class are as follows
+#'
+#' @slot name character representing name of model
+#' @slot output_names names of the compartments which are used by the
+#'     model.
+#' @slot parameter_names names of the initial conditions and parameters
+#'     which are used by the model.
+#' @slot parameters named list containing parameters of the model.
+#'     Initial values for each compartment, S0, E0, I0, R0.
+#'     Other parameters: a, b, c which represent the rates of changes
+#'     between the compartments.
+#' @slot n_age_categories number of age categories.
+#'
+#' @import deSolve
+#' @import glue
+#' @import ggplot2
+
+setClass('age_model',
+         #slots
+         slots = c(
+           name = 'character',
+           output_names = 'list',
+           parameter_names = 'list',
+           parameters = 'list',
+           n_age_categories = 'numeric'
+         ),
+
+         #prototypes for the slots, automatically set output and param
+         #names
+         prototype = list(
+           name = NA_character_,
+           output_names = list('S', 'E', 'I', 'R', 'Incidence'),
+           parameter_names = list('S0', 'E0', 'I0', 'R0', 'a', 'b', 'c'),
+           parameters = vector(mode = "list", length = 7),
+           n_age_categories = NA_real_
+
+         )
+)
+
+#' @describeIn Retrieves parameters for an age-structured simple SEIR model.
+#'
+#' @param object An object of the class age_model.
+setGeneric('get_parameters', function(object) standardGeneric('get_parameters'))
+
+setMethod(
+  'get_parameters', 'age_model',
+  function(object) object@parameters
+)
+
+#' @describeIn Sets parameters for an age-structured simple SEIR model.
+#'
+#' Default parameters a = 1, b = 1 and c = 1.
+#' If the initial conditions provided to do not sum to 1 or of different
+#' sizes compared to the number of age groups, an error is thrown.
+#'
+#' @param object An object of the class age_model.
+#' @param S0 initial fraction of the population that is susceptible
+#'           by age group.
+#' @param E0 initial fraction of the population that has been exposed
+#'           by age group.
+#' @param I0 initial fraction of the population that is infected
+#'           by age group.
+#' @param R0 initial fraction of the population that has recovered
+#'           by age group.
```

Comment:
Maybe it would be good to note what data type your function can handle here? Since you are dealing with age groups, you'll probably have vectors of some sort as input and I think it might be useful to a user to know which data types the code can handle?

Reply:
Yep, agreed.

Reply:
added text signally that data can be added as vector of fractions (1 element per age group).